File: 142.环形链表-ii.py
#
# @lc app=leetcode.cn id=142 lang=python3
#
# [142] 环形链表 II
#

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
from typing import Optional
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        d = set()
        while head is not None:
            if id(head) in d: return head
            d.add(id(head))
            head = head.next
        
# @lc code=end

File: 142.环形链表-ii.cpp
/*
 * @lc app=leetcode.cn id=142 lang=cpp
 *
 * [142] 环形链表 II
 */

#include<cstddef>
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include<unordered_set>
using namespace std;
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        unordered_set<ListNode*> d;
        while(head != NULL){
            if(d.count(head)) return head;
            d.insert(head);
            head = head->next;
        }
        return NULL;
    }
};
// @lc code=end

